### PR TITLE
Nrf52 qspi flash

### PR DESCRIPTION
**Pull request by hathach**
It is still work in progress but current code works with 8 MB MX25R6435F on pca10056. The read speed is ok, but the write is painfully slow (will improve later). I would like to have as many review/suggestion as possible. 

- added `flash_api` directory and move all the flash there
- flash_api.c implement generic read/write blocks API with page caching along with python glueing.
- flash_api use flash_hal_erase/read/program which are implemented by internal or qspi flash.
- flash hal will be dynamic binding later ( e.g switch from qspi to internal should we couldn't read/write to external flash in case of soldering issue etc ..)

#1049

### CODE REVIEW

**Comment by tannewt** [comment]
Write speed isn't critical to using CircuitPython. Yes, it'd be nice if it was faster but let's ignore it for now and focus on getting all of the APIs going.

I suggest filing a separate issue for it for the full 4.0.0 release milestone and moving onto other tasks for now.

**Comment by hathach** [comment]
@tannewt phew, I am trying to increase usb msc buffer size to 8KB but it doesn't help as well. probably need to dive deeper into the flash datasheet timing etc ... I will try to wrap it up now.

**Comment by dhalbert** [comment]
@hathach We'd like to make this implementation be as similar as possible to the atmel-samd QSPI and external flash implementation, so that we can factor out the common code as much as possible and only implement the lowest level routines in a chip-specific way. For instance, that code has parameterized chip support, and it already has caching logic. Did you try copying that impl? This impl is more different than we thought it would be.

**Comment by hathach** [comment]
@dhalbert many of the spi std command for flash is handled by nrfx qspi driver. Since nrf52840 has enough memory (4K) to cache the whole page. I just refactor both caching for internal + external flash giving it to only basic hal erase, read, program.  At anywhere you could use block API flash_read_blocks/flash_write_blocks instead of internal/external_flash_write_blocks. The flash_ will use either internal or external depending on the config. Those can be changed to function pointer for concurrent internal + external uses if needed.

**Comment by tannewt** [comment]
There is additional logic in the atmel-samd port that isn't in this one because it wasn't started from a copy. For example, the SPI flash chip [is detected](https://github.com/adafruit/circuitpython/blob/master/ports/atmel-samd/external_flash/external_flash.c#L184) rather than hard coded. Ideally this logic would be shared but for now it can be copied if that's easier.

Ideally you'd only need to replace the implementations [here](https://github.com/adafruit/circuitpython/blob/master/ports/atmel-samd/external_flash/qspi_flash.c) but it hasn't been tested yet.

**Comment by hathach** [comment]
I tried to enable high performance bit but couldn't get it written any faster. It is probably default mode already. Note: nrf52 SCK clock speed is only up to 32 Mhz. 
screenshot from 2018-09-16 00-20-58

I post an issue on Nordic devzone to see if there is anyone got more lucky than us.
https://devzone.nordicsemi.com/f/nordic-q-a/38570/write-speed-with-mx25r6435f-on-nrf52840dk

**Comment by hathach** [comment]
Nordic testing is a bit higher than us for write speed (different tool can be a bit different), but seem like write speed will be somewhat 1/10 of read speed. We may not improve it much with pca10056.

image

**Comment by dhalbert** [comment]
@hathach can you resolve the conflicts that have arisen? Thanks.

**Comment by hathach** [comment]
@dhalbert I am making more improvement while porting this for Arduino, I will clean up and update the flash code then resolve the conflict for this a bit later.

**Comment by dhalbert** [comment]
@hathach I stumbled on this while looking at the datasheet for something else. It says to use high drive strength to drive the QSPI pins:

[page 267, section 6.19.1]
>_2. To ensure stable operation, set the GPIO drive strength to “high drive”. See the GPIO — General
purpose input/output on page 141 chapter for details on how to configure GPIO drive strength._

File: ports/nrf/flash_api/flash_devices.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    uint16_t status_quad_enable;
+
+} qspi_flash_device_t;
+
+// All supported devices
+#define ALL_EXTERNAL_FLASH_DEVICES   GD25Q16C, MX25R6435F
+
+
+// Settings for the Gigadevice GD25Q16C 2MiB SPI flash.
+// Datasheet: http://www.gigadevice.com/wp-content/uploads/2017/12/DS-00086-GD25Q16C-Rev2.6.pdf
```
Comment (arturo182) [inline ports/nrf/flash_api/flash_devices.h]: This link returns 404 for me

File: ports/nrf/flash_api/flash_devices.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+#define GD25Q16C {\
+    .total_size = 2*1024*1024, \
+    .manufacturer_id = 0xc8, \
+    .device_id = 0x14, \
+    .status_quad_enable = (1 << 9)\
+}
+
+#define MX25R6435F {\
```
Comment (arturo182) [inline ports/nrf/flash_api/flash_devices.h]: Add a datasheet link as well?

**Comment by arturo182** [comment]
I'm trying to get this patch working on my board. I use a different flash chip but I added it to `flash_devices.h`, it seems to get detected properly because I get past the `if ( _flash_devices_arr[i].manufacturer_id == mfgr_id && _flash_devices_arr[i].device_id == dev_id ) {` check with proper values, however the MSD does not get mounted by the system (Linux Mint), trying to mount it manually gives me an error:
```
> sudo mount /dev/sdb /media/arturo182/usb
mount: /media/arturo182/usb: wrong fs type, bad option, bad superblock on /dev/sdb, missing codepage or helper program, or other error.
```

The CDC is properly detected and works just fine, but it also doesn't seem to be aware of the flash storage:
```
>>> import os
>>> os.listdir('.')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 19] Unsupported operation
```

Also, the chip i use is 128Mbit/16Mbyte and I specified it as so `.total_size = 16*1024*1024,` but when I connect the board I get this message `sd 4:0:0:0: [sdb] 32768 512-byte logical blocks: (16.8 MB/16.0 MiB)`, doesn't that mean that the system think there is more space than the chip has so either we might overwrite something or discard it at some boundary check? I think it's possible the `total_size` units are wrong.

**Comment by dhalbert** [comment]
@arturo182 What chip is it, and does it truly have 16MiB (binary megabytes), or is it smaller?

Also 16MB is just at the boundary between FAT12 and FAT16, and it's possible either Linux or FatFS is confused about just where that boundary is: one might be think it's FAT12 and the other might think it's FAT16.

**Comment by arturo182** [comment]
Ah yes, I should've written the model, it's  IS25LP128F, datasheet available here: http://www.issi.com/WW/pdf/25LP-WP128F.pdf
I also tried setting `total_size` to 8MB and that didn't work either, but I'm not sure if mismatching like that should work at all.

**Comment by hathach** [comment]
@aturo182 is the flash wired for quad mode ? If yes, Also please check the status registers mask in device config to make sure it is correct value to put flash device into quad mode.

**Comment by arturo182** [comment]
I use `(1 << 6)` for the quad enable value, I think it's correct according to the datasheet:
image

As for the design, I'm not sure what more is needed for quad mode wiring, I pretty much copied the PCA10056 except with a different chip, most qspi flash ICs seem to be pin compatible.

Datasheet:
image

My design:
image

@ladyada do you think it would be possible to post the nrf52840 feather schematic part with the QSPI flash so I could compare?

**Comment by tannewt** [comment]
@arturo182 I think most of the QSPI testing has been done on the Nordic dev boards. The feather exists but I'm not sure if anyone has tested it at all.

**Comment by arturo182** [comment]
I see, I thought @hathach was using some early nRF52840 feather prototypes to test the qspi :D

Looking at the PCA10056 schematic the pinout seems the same, I don't think there is any limit to which pins can be used as QSPI, so maybe I'll double check my soldering, but seems odd that it managed so read the chip id without a problem.
image

**Comment by hathach** [comment]
@arturo182 I only tested with pca10056, maybe you should start there to see if that works for you first and then moving to your custom board.

**Comment by arturo182** [comment]
Good idea, I'll try that :D

**Comment by arturo182** [comment]
I tested the code on PCA10056 and it works there so must be something with either the soldering, the chip or my gpio choice.

**Comment by tannewt** [comment]
FYI I'm moving the atmel-samd port to tiny usb and factoring out all of the usb stuff from the ports. This includes MSC so I plan on factoring out the spi flash chip support as well. I'll add support for QSPI on nRF based on this PR (thanks for it!) but will not merge it directly. Expect to see a PR for it next week sometime. Thanks!

**Pull request closed without merging**